File: strategy/typescript/src/strategy-oop.ts
type NotificationProvider = "email" | "sms" | "pushNotification";
type Message = string;
type AsyncTask = Promise<"success" | "failure">;

interface NotificationStrategy {
	notify: (message: Message) => AsyncTask;
}

type NotificationStrategies = Partial<
	Record<NotificationProvider, NotificationStrategy>
>;

type SendEmailParams = {
	emailAddress: string;
	subject: string;
	message: Message;
};
type SendEmailMethod = (params: SendEmailParams) => AsyncTask;
type EmailNotificationStrategyParams = {
	userEmail: string;
	subject: string;
	sendEmail: SendEmailMethod;
};

export class EmailNotificationStrategy implements NotificationStrategy {
	private userEmail: string;
	private subject: string;
	private sendEmail: SendEmailMethod;
	constructor({
		userEmail,
		subject,
		sendEmail,
	}: EmailNotificationStrategyParams) {
		this.subject = subject;
		this.userEmail = userEmail;
		this.sendEmail = sendEmail;
	}

	public notify = async (message: Message) => {
		const state = await this.sendEmail({
			emailAddress: this.userEmail,
			subject: this.subject,
			message: message,
		});

		return state;
	};
}

type SMSNotificationStrategyParams = {
	phoneNumber: string;
	sendSms: SendSMSMethod;
};
type SendSmsParams = { phoneNumber: string; message: Message };
type SendSMSMethod = ({ message, phoneNumber }: SendSmsParams) => AsyncTask;

export class SMSNotificationStrategy implements NotificationStrategy {
	private phoneNumber: string;
	private sendSms: SendSMSMethod;
	constructor({ phoneNumber, sendSms }: SMSNotificationStrategyParams) {
		this.phoneNumber = phoneNumber;
		this.sendSms = sendSms;
	}

	public notify = async (message: Message) => {
		const state = await this.sendSms({
			phoneNumber: this.phoneNumber,
			message,
		});

		return state;
	};
}

type UserId = string;
type SendPushNotificationParams = {
	userId: UserId;
	message: Message;
};
type SendPushNotificationMethod = (
	params: SendPushNotificationParams
) => AsyncTask;
type PushNotificationStrategyParams = {
	userId: UserId;
	sendPushNotification: SendPushNotificationMethod;
};

export class PushNotificationStrategy implements NotificationStrategy {
	private userId: UserId;
	private sendPushNotification: SendPushNotificationMethod;
	constructor({
		userId,
		sendPushNotification,
	}: PushNotificationStrategyParams) {
		this.userId = userId;
		this.sendPushNotification = sendPushNotification;
	}

	public notify = async (message: Message) => {
		const state = await this.sendPushNotification({
			userId: this.userId,
			message,
		});

		return state;
	};
}

export class Notifier {
	private strategies: NotificationStrategies;
	constructor({ strategies }: { strategies: NotificationStrategies }) {
		this.strategies = strategies;
	}

	public notify = async ({
		provider,
		message,
	}: {
		provider: NotificationProvider;
		message: Message;
	}) => {
		const strategy = this.strategies[provider];

		if (!strategy) {
			throw new Error(`Notification strategy for ${provider} is not provided.`);
		}

		return await strategy.notify(message);
	};
}

File: strategy/python/strategy_oop.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Literal, Union, Awaitable, TypedDict, Callable, Dict, Optional

NotificationProvider = Union[
    Literal["email"], Literal["sms"], Literal["push_notification"]
]
Message = str
AsyncTask = Awaitable[Literal["success", "failure"]]


class NotificationStrategy(ABC):
    @abstractmethod
    def notify(self, message: Message) -> AsyncTask:
        pass


class SendEmailParams(TypedDict):
    email_address: str
    subject: str
    message: Message


SendEmailMethod = Callable[[SendEmailParams], AsyncTask]


class EmailNotificationStrategyParams(TypedDict):
    user_email: str
    subject: str
    send_email: SendEmailMethod


class EmailNotificationStrategy(NotificationStrategy):
    def __init__(self, params: EmailNotificationStrategyParams) -> None:
        self.user_email = params["user_email"]
        self.subject = params["subject"]
        self.send_email = params["send_email"]

    async def notify(self, message: str):
        email_params: SendEmailParams = {
            "email_address": self.user_email,
            "subject": self.subject,
            "message": message,
        }
        return await self.send_email(email_params)


NotificationStrategies = Dict[NotificationProvider, Optional[NotificationStrategy]]


@dataclass
class NotifyParams:
    provider: NotificationProvider
    message: Message


class Notifier:
    def __init__(self, strategies: NotificationStrategies) -> None:
        self.strategies = strategies

    async def notify(self, params: NotifyParams):
        strategy = self.strategies[params.provider]

        if not (strategy):
            raise Exception("Notification strategy for ${provider} is not provided.")

        return await strategy.notify(params.message)
